src/engine.rs

```diff
@@ -422,8 +308,16 @@ impl Engine {
     #[cfg(feature = "testnet")]
     /// Credits the address with 10 coins from the faucet.
     pub fn credit(&mut self, address: &Address) -> EngineResult<()> {
-        let new_bal = Self::check_increase_balance(address, &U256::from(10))?;
-        Self::set_balance(address, &new_bal);
+        use crate::prelude::Add;
+
+        let balance = Self::get_balance(address);
+        // Saturating adds are intentional
+        let new_balance = balance.saturating_add(U256::one());
+
+        let account_nonce = Self::get_nonce(address);
+        account_nonce.saturating_add(U256::one());
```

Comment:
Should be done in `call` no?

Reply:
`credit` is only called from `make_it_rain`, which is only callable via a Near transaction, and the account making that Near transaction has no intrinsic affiliation with the address being credited. It's like how in a transfer you increment the nonce of the source account, not the destination account. In this case the "source account" is external to the VM, so there's no nonce to change.

Reply:
Ok, I'll take a look tomorrow morning.

Reply:
Note that calls (such as `make_it_rain`) via the NEAR function call interface are considered "virtual" EVM transactions originating from a Keccak256-derivation of the NEAR account ID that signed the transaction. These virtual EVM transactions are exposed as such via the RPC interface. That means that their side effects must be indistinguishable from signed, native EVM transactions via `submit`. So, we do need to maintain nonces for addresses that map to NEAR account IDs.

Reply:
>  "virtual" EVM transactions originating from a Keccak256-derivation of the NEAR account ID 

Currently that is not true for `make_it_rain`. I would be fine with incrementing the nonce of that account. But we should not increment the nonce of the account passed as an argument (i.e. read via `read_input`) because there is no validation there and so anyone could increment anyone else's nonce (on testnet).

Reply:
@birchmd Good point. Given which, the thing to do here is to remove the address argument from `make_it_rain` and instead rely solely on `predecessor_address()`. That does limit the utility somewhat compared to what we initially figured, but ultimately only means that to transfer funds to an arbitrary EVM address on TestNet takes two calls (`make_it_rain` and `transfer`, the latter which we are still lacking) instead of one.

Reply:
I think it would be ok for `make_it_rain` to still take the destination account as an argument, but we should run that transaction as if it were started by `predecessor_address()`. So something like

```rust
fn make_it_rain() {
    let input = sdk::read_input();
    let dest_address = Address::from_slice(&input);
    let source_address = predecessor_address();
    let mut engine = Engine::new(source_address);
    
    engine.increment_nonce(source_address);
    engine.credit(dest_address).sdk_unwrap();
}
```

Reply:
I like that approach @birchmd. I just wrote it up for that exact recommendation. So, going through all the messages, I should've nailed everything.

A good check on the oversight. Makes no sense to increment the nonce of the receiver. However, does it make sense to really increment at all? There is nothing linking the transaction at this nonce increment to the `predecessor_address`? I want to make it indistinguishable but, these are just are free money on the test net with some balance change hacks so to speak.

Reply:
What I mean is, we have nothing that really says "We did this action from this address, to that address". We are just ghost incrementing the nonce. So I'm unsure, on second glance, if its even worth it now.

Reply:
For consistency sake though, sure its fine. I added a commit on that now.

Reply:
@joshuajbouw Nonces must be maintained. As said, these mutations will be exposed as virtual transactions.